lib/client-api/src/story_store.ts

```diff
@@ -383,8 +383,32 @@ export default class StoryStore {
     const loaders = [...this._globalMetadata.loaders, ...kindMetadata.loaders, ...storyLoaders];
 
     const finalStoryFn = (context: StoryContext) => {
-      const { passArgsFirst = true } = context.parameters;
-      return passArgsFirst ? (original as ArgsStoryFn)(context.args, context) : original(context);
+      const { args, argTypes, parameters } = context;
+      if (
+        argTypes &&
+        Object.values(argTypes).some(({ control }) => {
+          if (!control?.options) return false;
+          return Object.values(control.options).some((v: any) => v && typeof v === 'object');
+        })
+      ) {
+        once.warn(
+          'Objects cannot be used as values in control options. Use a `mapping` instead.\n\nMore info: https://storybook.js.org/docs/react/writing-stories/args#mapping-to-complex-arg-values'
+        );
+      }
+
+      const { passArgsFirst = true } = parameters;
+      if (args) {
+        const mapped = {
+          ...context,
+          args: Object.entries(args).reduce((acc, [key, val]) => {
+            const { mapping } = argTypes?.[key] || {};
+            acc[key] = mapping && val in mapping ? mapping[val] : val;
+            return acc;
+          }, {} as Args),
+        };
+        return passArgsFirst ? (original as ArgsStoryFn)(mapped.args, mapped) : original(mapped);
+      }
+      return passArgsFirst ? (original as ArgsStoryFn)(args, context) : original(context);
```

Comment:
Can't we just default args to `{}`? I'm not sure it is even possible for them to be undefined.

Reply:
Sure, I just didn't want to accidentally break anything.